src/emitters/sky.cpp

```diff
@@ -0,0 +1,203 @@
+#include <mitsuba/core/bsphere.h>
+#include <mitsuba/core/fresolver.h>
+#include <mitsuba/core/plugin.h>
+#include <mitsuba/core/properties.h>
+#include <mitsuba/core/warp.h>
+#include <mitsuba/render/emitter.h>
+#include <mitsuba/render/scene.h>
+#include <mitsuba/render/texture.h>
+#include "sunsky/sunmodel.h"
+#include "sunsky/skymodel.h"
+
+NAMESPACE_BEGIN(mitsuba)
+
+/**!
+
+.. _emitter-sky:
+
+Skylight emitter (:monosp:`sky`)
+-------------------------------------------------
+
+.. pluginparameters::
+
+ * - radiance
+   - |spectrum|
+   - Specifies the emitted radiance in units of power per unit area per unit steradian.
+     (Default: :ref:`emitter-d65`)
+
+This plugin implements a skylight emitter, which surrounds
+the scene and radiates diffuse illumination towards it. This is often
+a good default light source when the goal is to visualize some loaded
+geometry that uses basic (e.g. diffuse) materials.
+
+ */
+
+template <typename Float, typename Spectrum>
+class SkyEmitter final : public Emitter<Float, Spectrum> {
+public:
+    MTS_IMPORT_BASE(Emitter, m_flags)
+    MTS_IMPORT_TYPES(Scene, Shape, Texture)
+
+    SkyEmitter(const Properties &props) : Base(props) {
+        /* Until `set_scene` is called, we have no information
+           about the scene and default to the unit bounding sphere. */
+        m_bsphere = ScalarBoundingSphere3f(ScalarPoint3f(0.f), 1.f);
+
+        m_flags = +EmitterFlags::Infinite;
+
+        m_scale = props.float_("scale", 1.0f);
+        m_turbidity = props.float_("turbidity", 3.0f);
+        m_stretch = props.float_("stretch", 1.0f);
+        m_resolution = props.int_("resolution", 512);
+        m_albedo = props.texture<Texture>("albedo", 0.2f);
+        m_sun = compute_sun_coordinates<Float>(props);
+        m_extend = props.bool_("extend", false);
+
+        if (m_turbidity < 1 || m_turbidity > 10)
+            Log(Error, "The turbidity parameter must be in the range[1,10]!");
+        if (m_stretch < 1 || m_stretch > 2)
+            Log(Error, "The stretch parameter must be in the range [1,2]!");
+        for (size_t i = 0; i < Spectrum::Size; ++i) {
+            if (m_albedo[i] < 0 || m_albedo[i] > 1)
+                Log(Error, "The albedo parameter must be in the range [0.1]");
+        }
+
+        Float sun_elevation = 0.5f * math::Pi<Float> - m_sun.elevation;
+
+        if (sun_elevation < 0)
+            Log(Error, "The sun is below the horizon -- this is not supported by the sky model.");
+
+        for (size_t i = 0; i < Spectrum::Size; i++) {
+            if constexpr (Spectrum::Size == 3)
+                m_state[i] = arhosek_rgb_skymodelstate_alloc_init(
```

Comment:
Since the model states are allocated here, they should also be freed in the destructor (`arhosekskymodelstate_free`, and corresponding in the spectral case).